libs/libarchfpga/src/read_xml_arch_file.cpp

```diff
@@ -2754,20 +2803,163 @@ static void ProcessComplexBlocks(pugi::xml_node Node,
         Cur = get_single_child(CurType, "fc", loc_data, OPTIONAL);
         Process_Fc(Cur, Type, arch.Segments, arch_def_fc, loc_data);
 
-        //Load switchblock type and location overrides
+        /* Load switchblock type and location overrides */
         Cur = get_single_child(CurType, "switchblock_locations", loc_data, OPTIONAL);
         ProcessSwitchblockLocations(Cur, Type, arch, loc_data);
 
-		Type->index = i;
-
-		/* Type fully read */
-		++i;
+		/* Load possible modes (pb_types which are compatible with the current tile) */
+		Cur = get_single_child(CurType, "equivalent_tiles", loc_data, OPTIONAL);
+		if(Cur) {
+			ProcessTileExtraModes(Cur, Type, TypeMap, loc_data);
+		}
 
 		/* Free this node and get its next sibling node */
 		CurType = CurType.next_sibling (CurType.name());
+	}
+}
+
+/* Processes the equivalent tiles defined in the XML arch definition
+ * <tiles>
+ *   <tile name="LAB">
+ *     <mode name="MLAB">
+ *       <map .../>
+ *       <map .../>
+ *       <map .../>
+ *     </mode>
+ *   </tile>
+ * </tiles>
+ *
+ * In particular this function parses the `modes` (if they exist) of each tile
+ * and adds the equivalent tile information to the t_type_descriptor relative to
+ * the current tile.
+ * It populates the following t_type_descriptor members:
+ *     - num_equivalent_tiles;
+ *     - equivalent_tiles.
+ */
+static void ProcessTileExtraModes(pugi::xml_node Node,
+		t_type_descriptor *Type,
+		std::unordered_map<std::string, t_type_descriptor *> TypeMap,
+		const pugiutil::loc_data& loc_data) {
+	pugi::xml_node CurType;
+
+	Type->num_equivalent_tiles = count_children(Node, "mode", loc_data);
+	Type->equivalent_tiles = (t_type_descriptor **) vtr::malloc(Type->num_equivalent_tiles * sizeof(t_type_descriptor *));
```

Comment:
Please just use std::vector

Reply:
See https://github.com/verilog-to-routing/vtr-verilog-to-routing/issues/506

Reply:
Done, I have converted all the new members related to equivalent tiles (e.g. `pin_mapping` as well) to `unordered_maps`.
I have used `unordered_maps` as I need to access to elements by indices and vectors are not too suited for index access.